dnf5: Support for groups installation

m-blaha opened the pull request.

m-blaha committed:
libdnf-cli: Allow to construct StringArgumentList with nargs

m-blaha committed:
dnf5/group: GroupSpecArguments with nargs

m-blaha committed:
dnf5/group: Install command

m-blaha committed:
libdnf-cli: Add groups to transaction table

m-blaha committed:
libdnf/base: Goal::Impl::add_install_to_goal returns id_queue

m-blaha committed:
dnfdaemon: Prepare for groups in transactions

m-blaha committed:
libdnf/comps: Unify PackageType and CompsPackageType

I guess we do not need two different enums for the same thing.

m-blaha committed:
libdnf: Implement group installation

j-mracek commented on include/libdnf/base/goal_elements.hpp:
```diff
@@ -95,6 +96,21 @@ enum class GoalAction {
+    bool with_group_id{false};
+    bool with_group_name{false};
+
+    bool with_optional{false};
```
Groups uses I thing four types of packages - `mandatory`, `default`, `optional`, `conditional`. In DNF on API it was possible to handle each of those types separatelly or all together therefore I am not sure whether handling only optional is sufficient. I know that many applications reduce the behavior only to two groups - default dnf set plus optional.

m-blaha replied:
Right. I'll change the `bool with_optional` to `libdnf::comps::PackageType package_types` defaulting to `MANDATORY | DEFAULT | CONDITIONAL`

j-mracek commented on libdnf/base/goal.cpp:
```diff
@@ -905,6 +936,62 @@ void Goal::Impl::add_up_down_distrosync_to_goal(
+            auto pkg_name = pkg.get_name();
+            auto pkg_condition = pkg.get_condition();
+            if (pkg_condition.empty()) {
+                auto [pkg_problem, pkg_queue] = add_install_to_goal(transaction, pkg_name, pkg_settings);
```
I was thinking that to use `add_install_to_goal()` is incorrect but right now I thing it is quite clever because of multilib setting. Only one thing that may be can be improved are reports of missing packages. May be it would be nice to report missing packages from groups by another way or to allow identification of reports generated from user requested job (package install) or as an consequence of other request (installation of package required by a group).

m-blaha replied:
I'll add new action `INSTALL_BY_GROUP`.

j-mracek commented on libdnf/base/goal.cpp:
```diff
@@ -905,6 +936,62 @@ void Goal::Impl::add_up_down_distrosync_to_goal(
+    group_query.filter_groupid(spec, sack::QueryCmp::IGLOB);
+    group_query |= group_query_names;
+    if (group_query.empty()) {
+        auto problem = transaction.p_impl->report_not_found(GoalAction::INSTALL, spec, GoalJobSettings(), strict);
```
It will be quite difficult to report group problem in current goal log. I believe that it is a right place but the place is not optimized to accept GroupJobSettings. May be we can discuss it.

j-mracek commented on libdnf/base/goal.cpp:
```diff
@@ -905,6 +936,62 @@ void Goal::Impl::add_up_down_distrosync_to_goal(
+    comps::GroupQuery group_query_names(group_query);
+    group_query_names.filter_name(spec, sack::QueryCmp::IGLOB);
+    group_query.filter_groupid(spec, sack::QueryCmp::IGLOB);
+    group_query |= group_query_names;
```
GroupJobSetting contains icase, group id and group name setting but it is not used. Probably it was intended to be used here.

m-blaha replied:
Yes, I've just added a commit that implements it.

j-mracek commented on libdnf/base/goal.cpp:
```diff
@@ -905,6 +936,62 @@ void Goal::Impl::add_up_down_distrosync_to_goal(
+                rpm_goal.add_group_installed(pkg_queue);
+                packages_res |= pkg_problem;
+            } else {
+                add_provide_install_to_goal(fmt::format("({} if {})", pkg_name, pkg_condition), pkg_settings);
```
What about to move `add_provide_install_to_goal` to `if (!query.empty()) {` and then also create `else` to report that group package is missing or we can also check the condition and only in case that condition is met by any package then we can create the report. What do you think?

m-blaha replied:
I'm not sure about reporting a missing package. At this stage we do not know whether the condition is met for the transaction and reporting a missing package could be misleading in case the condition is not met.

j-mracek commented on libdnf/base/goal_elements.cpp:
```diff
@@ -25,6 +25,29 @@ along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
 namespace libdnf {
 
 
+bool GroupJobSettings::resolve_strict(const libdnf::ConfigMain & cfg_main) {
```
[optional] What about to create a template for `GroupJobSettings::resolve_strict(const libdnf::ConfigMain & cfg_main)` and `GoalJobSettings::resolve_strict(const libdnf::ConfigMain & cfg_main)`. It looks like that they are identical.

j-mracek commented on libdnf/rpm/solv/goal_private.cpp:
```diff
@@ -632,8 +632,22 @@ transaction::TransactionItemReason GoalPrivate::get_reason(Id id) {
+            }
+        }
+        // explicitely group-installed
+        for (const auto & ui_id : group_installed) {
```
It looks like a good place to handle it, but I afraid about side effects. I've been playing with `dnf shell` and I combined remove and install request and I was able to get reinstall operation (it can happen as well with group install and remove). Then the reason for inbound package will be correct but the reason for removed packages will be also group. The installation of group can also trigger upgrade (package installed, new version available with `--best`). It means that replace package will have also reason group. I am really not sure whether it is a problem or not. @lukash Please what is your opinion?

lukash replied:
(sorry it's a bit of a brain-dump)

I am not sure. First, I'm operating under the assumption that we should record all packages from a group that would be installed by it, but are already installed, as "been installed from the group". That means putting them on the list of installed packages of that group in `groups.toml`.

At the same time they have a reason recorded in `packages.toml`, where I'm not sure if we should use the `GROUP` reason. It would appear we should do it for "consistency" (of the `reason` attribute itself). OTOH, when that group is removed and we should change the `reason` in `packages.toml` to whatever it was before (I guess `EXTERNAL_USER`, `DEPENDENCY` and `WEAK_DEPENDENCY` are the possible values here), I have no idea how to retrieve what was there before without recursing into the history database, which is one of the core design concepts that we don't do that.

I don't even know what should be the reason for the `REMOVE` action, because there the semantics shift (the `reason` becomes the reason of why the package was removed, as opposed to why the package is installed). I guess for printing out the transaction table having `GROUP` there is good?

For upgrades etc., I guess the reason `GROUP` is good, as we do want to mark the package as installed by the group. But then, we need that for even packages that aren't even being upgraded (they are just already installed).

And, on top of all that, we don't only need the `reason` for the packages installed from a group, we also need the names of groups that they were installed by, so that we can record it in both system state and the history db.

j-mracek commented on libdnf/rpm/solv/goal_private.cpp:
```diff
@@ -632,8 +632,22 @@ transaction::TransactionItemReason GoalPrivate::get_reason(Id id) {
-         solver_ruleclass(libsolv_solver, info) == SOLVER_RULE_BEST))
+         solver_ruleclass(libsolv_solver, info) == SOLVER_RULE_BEST)) {
+        // explicitely user-installed
+        for (const auto & ui_id : user_installed) {
```
Sequential search is quite slow (in comparison to ordered list, set or map). For the transaction with a lot of package to be installed it can be measurable. Think about anaconda - there is a lot of packages installed by groups. The scan is also applied for removed|replaced packages and it is completely unneeded and may be harmful.

m-blaha replied:
I will change IdQueue to SolvMap for better performance.

j-mracek commented on libdnf/base/transaction.cpp:
```diff
@@ -453,6 +463,21 @@ Transaction::TransactionRunResult Transaction::Impl::run(
+        for (const auto & pkg : packages) {
+            installed_names.emplace(pkg.get_package().get_name());
+        }
+        db_transaction.fill_transaction_groups(groups, installed_names);
```
I am sorry but the calculation of installed names and using them for transaction groups is somehow strange to me but I am sure it is functional. In goal install and goal_private are kept IDs of packages requested by group to calculate a reason. Also group install can store packages names after calling and verifying of rc==NO_PROBLEM from add_install_to_goal().

j-mracek replied:
May be I overlooked something, but how the code handles situation when package is already installed and then the package is installed as a part of group but rpm is not in transaction. I am sure that it will be recorded in db_transaction.fill_transaction_groups() but what about toml for rpms? @lukash Please what's your plans?

lukash replied:
I think I mostly covered this in my other comment: https://github.com/rpm-software-management/libdnf/pull/1563#discussion_r932012931 Though I'm not sure we agree on that and it has a number of loose ends.

j-mracek commented on libdnf/transaction/transaction.cpp:
```diff
@@ -206,6 +205,31 @@ void Transaction::fill_transaction_packages(
+            new_grp_pkg.set_package_type(group_package.get_type());
+            new_grp_pkg.set_installed(installed_names.contains(name) ? true : false);
+            package_types |= group_package.get_type();
+        }
```
May be it will sound strange but what about installation of a group without optional packages but I believe that the code will also mark all installed packages from optional group as group installed (the same problem can be achieved using multiple approaches). May be it would be nice to track more precisely what was installed by particular group. Or to use lass precise model for storing data. Somehow I feel that it is something in the middle. But it can be resolved by TODO.

m-blaha committed:
Replace GroupJobSetting.with_optional with package_types

This enables installation of group with packages of any types. Current
behaviour always installs default, mandatory, and conditional packages
and only enables user to add optional packages.

m-blaha committed:
comps/group: Allow empty query initialization

m-blaha committed:
libdnf/goal: Use GroupJobSettings flags to select group

By default the group spec is searched using case-insensitive match
against group ID and group name.

dmach commented on include/libdnf/base/goal_elements.hpp:
```diff
@@ -95,6 +97,23 @@ enum class GoalAction {
+struct GroupJobSettings {
+    bool ignore_case{true};
+    bool with_group_id{true};
+    bool with_group_name{true};
```
Don't forget that dnf4 supports also translated group names.
IMO using either untranslated or translated group name as an install spec is a very bad idea.
I'd prefer to set this to false by default: `bool with_group_name{false};` or even better - disable the functionality completely. Installations should be done by `group id`.

m-blaha replied:
Right, in dnf4 the `group_id` was somehow "hidden" and `group list` command required special option `--ids` to reveal it to the user. Given that `dnf5 group list` now directly presents group ID, moving from group name  to id as primarily used identifier makes sense.

m-blaha commented:
Replaced by PR https://github.com/rpm-software-management/dnf5/pull/7 in dnf5 repository.
